src/rgw/rgw_reshard.cc

```diff
@@ -222,6 +244,32 @@ void RGWBucketReshard::unlock_bucket()
   }
 }
 
+int RGWBucketReshard::renew_lock_bucket(const Clock::time_point& now)
+{
+  // assume outer locks have timespans at least the size of ours, so
+  // can call inside conditional
+  if (renew_locks_callback) {
+    int ret = renew_locks_callback(now);
+    if (ret < 0) {
+      return ret;
+    }
+  }
+
+  reshard_lock.set_must_renew(true);
+  int ret = reshard_lock.lock_exclusive(&store->reshard_pool_ctx, reshard_oid);
```

Comment:
one thing that stands out here is that we're taking locks on ephemeral objects in this reshard pool, that use the same names as the bucket instance objects in the metadata pool. so after reshard completes and eventually cleans up the bucket index shards and bucket instance metadata, these ephemeral objects will remain in the reshard pool forever

i wonder if we could avoid taking these extra per-bucket locks (thus creating these ephemeral lock objects), and only rely on the per-log-shard locks to get the same mutual exclusion. from what i can tell, the per-bucket locks are only really necessary to protect against radosgw-admin commands - could we change those to lock the associated log shard instead?

Reply:
that's a great catch;  that pattern (creating "ephemeral" objects or omap entries for some process and never removing them) is one we've been trying to eliminate

Reply:
The logshard objects do not exist when you issue an immediate reshard command, as in 
```
bin/radosgw-admin bucket reshard --bucket=${bucket} --num-shards=15
```
They only exist during dynamic resharding and "scheduled" resharding, as in:
```
bin/radosgw-admin reshard add --bucket=${bucket} --num-shards=20
bin/radosgw-admin reshard process
```
Which pool are these ephemeral objects in? When I list each of the pools on my cluster {
.rgw.root, default.rgw.control, default.rgw.meta, default.rgw.log, default.rgw.buckets.index, default.rgw.buckets.data}, I cannot see these ephemeral objects.

Reply:
the reshard_pool defaults to `.rgw.log:reshard`, or the 'reshard' namespace of the .rgw.log pool. try listing with `rados ls -p .rgw.log -N reshard`

Reply:
So we discussed having locks on ephemeral objects clean themselves up after they're unlocked. We further discussed that this behavior should only be allowed with exclusive locks (and not shared locks).

Implementation does not look very difficult. There are two basic approaches. My less favorite is passing flags with the unlock operation to indicate deletion should occur. For normal behavior this would requiring passing a 0 to indicate no flags (given that ceph coding standards disfavors default argument values).

My preferred approach is to mark the lock as ephemeral upon creation, so all existing unlocks do not have to indicate deletion. There would essentially be 3 kinds of locks -- SHARED, EXCLUSIVE, EXCLUSIVE_EPHEMERAL.

@mattbenjamin, @cbodley Thoughts?

Reply:
i guess that second/preferred approach is the only one that can handle racing renewals.. consider the case where reshard A starts but some operation stalls and its lock expires. another reshard B comes in and gets the lock, successfully reshards, and removes the lock object. reshard A eventually tries to renew its lock (which recreates the lock object) and fails - so this failed renewal request would also need to remove the object